src/azul/service/responseobjects/cart_item_manager.py

```diff
@@ -42,29 +42,36 @@ def create_cart(self, user_id, cart_name, default):
                                       key_conditions=query_dict,
                                       index_name='UserCartNameIndex') > 0:
             raise DuplicateItemError(f'Cart `{cart_name}` already exists')
-
         if default:
-            if self.dynamo_accessor.count(table_name=config.dynamo_cart_table_name,
-                                          key_conditions={'UserId': user_id},
-                                          filters={'DefaultCart': True},
-                                          index_name='UserIndex') > 0:
+            user = self.user_service.get(user_id)
+            if user['DefaultCartId'] is not None:
                 raise DuplicateItemError('Default cart already exists')
-
-        cart_id = 'default' if default else str(uuid.uuid4())
+        cart_id = str(uuid.uuid4())
         self.dynamo_accessor.insert_item(config.dynamo_cart_table_name,
-                                         item={'CartId': cart_id, 'DefaultCart': default, **query_dict})
+                                         item={'CartId': cart_id, **query_dict})
+        if default:
+            self.user_service.update(user_id, default_cart_id=cart_id)
```

Comment:
Consider making this update [conditional](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.ConditionExpressions.html#Expressions.ConditionExpressions.SimpleComparisons). It would be more efficient (because you can eliminate the user retrieval) and eliminate a race. There is a risk that the actual cart creation below fails but I think that possibility is far more unlikely than that of two concurrent requests for the creation of a default cart. I can elaborate why if needed.

Reply:
Currently, `UserService.get` is technically `get_or_create`. The first user retrieval serves two purposes: create the user object if not already existed, and check whether the user already has the default cart. When there are two concurrent requests, one of them should fail as both tries to create two objects with the same hash key.

I would like to discuss with you on the implementation tomorrow after the standup to see what we can do to eliminate the race condition.

Reply:
The refactored code to use `ConditionExpression` and reduce the chance of retrieving user to minimize race condition will be ready by the next review request.